bundles/org.openhab.binding.mycroft/src/main/java/org/openhab/binding/mycroft/internal/MycroftHandler.java

```diff
@@ -0,0 +1,234 @@
+        registerChannel(new VolumeChannel(this));
+        registerChannel(new MuteChannel(this));
+        registerChannel(new DuckChannel(this));
+        registerChannel(new SpeakChannel(this));
+        registerChannel(new AudioPlayerChannel(this));
+        registerChannel(new UtteranceChannel(this));
+
+        final Channel fullMessageChannel = getThing().getChannel(MycroftBindingConstants.FULL_MESSAGE_CHANNEL);
+        @SuppressWarnings("null") // cannot be null
+        String messageTypesProperty = (String) fullMessageChannel.getConfiguration()
+                .get(MycroftBindingConstants.FULL_MESSAGE_CHANNEL_MESSAGE_TYPE_PROPERTY);
+
+        registerChannel(new FullMessageChannel(this, messageTypesProperty));
+
+        checkLinkedChannelsAndRegisterMessageListeners();
+    }
+
+    private void checkLinkedChannelsAndRegisterMessageListeners() {
+        for (Entry<ChannelUID, MycroftChannel<?>> channelEntry : mycroftChannels.entrySet()) {
+            ChannelUID uid = channelEntry.getKey();
+            MycroftChannel<?> channel = channelEntry.getValue();
+            if (isLinked(uid)) {
+                channel.registerListeners();
+            } else {
+                channel.unregisterListeners();
+            }
+        }
+    }
+
+    @Override
+    public void channelLinked(ChannelUID channelUID) {
+        checkLinkedChannelsAndRegisterMessageListeners();
+    }
+
+    @Override
+    public void channelUnlinked(ChannelUID channelUID) {
+        checkLinkedChannelsAndRegisterMessageListeners();
+    }
+
+    private void registerChannel(MycroftChannel<?> channel) {
+        mycroftChannels.put(channel.getChannelUID(), channel);
+    }
+
+    public void registerMessageListener(MessageType messageType, MycroftMessageListener<?> listener) {
+        this.connection.registerListener(messageType, listener);
+    }
+
+    public void unregisterMessageListener(MessageType messageType, MycroftMessageListener<?> listener) {
+        this.connection.unregisterListener(messageType, listener);
+    }
+
+    @Override
+    public void connectionEstablished() {
+        logger.debug("Mycroft thing {} is online", thing.getUID());
+        updateStatus(ThingStatus.ONLINE);
+    }
+
+    @Override
+    public void connectionLost(String reason) {
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, reason);
+    }
+
+    @Override
+    public void dispose() {
+        thingDisposing = true;
+        stopTimer();
+        connection.close();
+    }
+
+    public <T extends State> void updateMyChannel(MycroftChannel<T> mycroftChannel, T state) {
+        updateState(mycroftChannel.getChannelUID(), state);
+    }
+
+    public boolean sendMessage(BaseMessage message) {
+        try {
+            connection.sendMessage(message);
+            return true;
+        } catch (IOException e) {
+            logger.warn("Cannot send message of type {}, for reason {}", message.getClass().getName(), e.getMessage());
```

Comment:
What can cause the IOException? Not that you should only log warnings, if this is considered a bug in the code that should be reported as an issue. If it can happen because of network issues, you must instead log debug only and set the Thing status accordingly.

Reply:
Network issue will cause IO exception.
So OK for debug, done.